app.go

```diff
@@ -453,28 +464,92 @@ func ValidateApp(opts ...Option) error {
 // registered via Invoke options. See the documentation of the App struct for
 // details on the application's initialization, startup, and shutdown logic.
 func New(opts ...Option) *App {
-	logger := fxlog.DefaultLogger(os.Stderr)
-
 	app := &App{
-		log:          logger,
+		log:          new(logBuffer),
 		startTimeout: DefaultTimeout,
 		stopTimeout:  DefaultTimeout,
 	}
 
 	for _, opt := range opts {
 		opt.apply(app)
 	}
-	app.lifecycle = &lifecycleWrapper{lifecycle.New(app.log)}
 
 	app.container = dig.New(
 		dig.DeferAcyclicVerification(),
 		dig.DryRun(app.validate),
 	)
+	var connectLogger func(logger fxlog.Logger) error
+	if app.fxlogConstructor != nil {
+		// Since user supplied a custom logger, use a buffered logger to hold
+		// all messages until user supplied logger is instantiated. Then we flush
+		// those messages after fully constructing the custom logger.
+		var buffer logBuffer
+		app.log = &buffer
+		connectLogger = buffer.Connect // func(fxlog.Logger) error
+	} else {
+		// If user hasn't supplied a custom implementation of fxlog.Logger
+		// then use default one.
+		app.log = fxlog.DefaultLogger(os.Stderr)
+	}
 
 	for _, p := range app.provides {
 		app.provide(p)
 	}
 
+	if app.err != nil {
+		return app
+	}
+
+	// If WithLogger and Printer are both provided, WithLogger takes precedence.
+	// If user supplied an fxlog constructor, we need to have dig's container
+	// initialize it and grab all of arguments from the graph and then return.
+	if app.fxlogConstructor != nil {
+		// Here we add the custom logger into the container.
```

Comment:
This if block is massive. Can we pull it into a method responsible for building the logger and nothing else?

```
if app.fxlogConstructor != nil {
  log, err := app.buildCustomLogger()  // better name TBD
  if err != nil {
    // failed to build custom logger. Flush to stderr.
    connectLogger(fxlog.DefaultLogger(os.Stderr))
    app.err = err
    return app
  }
  connectLogger(log)
}
```